### ISSUE TITLE
Walrus operator + any comprehension confuses mypy?

### ISSUE THREAD

**Issue by shartzog** [python/mypy#15335]
mypy returns an unexpected used-before-def error when a walrus operator is used to capture the matching element in an `any` comprehension.

Given test_mypy.py containing:
```python
test = ['a', 'b', 'c']
if any((_t := itm) == 'c' for itm in test):
    print(f'{_t=}')
else:
    _t = None
print(_t)
```

mypy produces:
```cmd
>mypy test_mypy.py
test_mypy.py:3: error: Name "_t" is used before definition  [used-before-def]
Found 1 error in 1 file (checked 1 source file)
```

Let me know if this is captured elsewhere in list of walrus operator issues, or if it's not a problem for some reason I'm not seeing. Thanks!

mypy==1.3.0
python==3.10.4
no cmd line options outside of the test file name.

**Comment by TeamSpen210** [comment]
The issue here is that it's only safe because of the specific behaviour of `any()`. Since generators only actually execute when iterated over, if `any()` didn't iterate the generator `_t` would never be assigned. `any()` always does iterate it, but that might not be true for an arbitrary function. So this would require special casing.

**Comment by shartzog** [comment]
FYI, the easiest workaround here is a pre-emptive type annotation if anyone else is annoyed by this ;)

```python3
test = ['a', 'b', 'c']
_t: str  # <-- this prevents the false positive
if any((_t := itm) == 'c' for itm in test):
    print(f'{_t=}')
else:
    _t = None
print(_t)
```
